src/visit.rs

```diff
@@ -0,0 +1,44 @@
+use squalid::OptionExt;
+use tree_sitter_grep::tree_sitter::{Node, Tree};
+
+pub trait TreeEnterLeaveVisitor<'a> {
+    fn enter_node(&mut self, node: Node<'a>);
+    fn leave_node(&mut self, node: Node<'a>);
+}
+
+pub fn walk_tree<'a>(tree: &'a Tree, visitor: &mut impl TreeEnterLeaveVisitor<'a>) {
+    let mut node_stack: Vec<Node<'a>> = Default::default();
+    let mut cursor = tree.walk();
+    'outer: loop {
+        let node = cursor.node();
+        while node_stack
+            .last()
+            .matches(|&last| node.end_byte() > last.end_byte())
+        {
+            // trace!(target: "visit", ?node, "leaving node");
```

Comment:
These were throwing off the timing shown in the tracing-chrome traces a fair amount so maybe leaving them commented-out is a good default